protocol/proto/encode.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+
+	"github.com/m3db/m3x/pool"
+)
+
+type encoder struct {
+	w          *bufio.Writer
+	sizeBuffer []byte
+	dataBuffer []byte
+	bytesPool  pool.BytesPool
+}
+
+// NewEncoder creates a new encoder.
+func NewEncoder(w io.Writer, opts EncodeDecoderOptions) Encoder {
+	return newEncoder(w, opts)
+}
+
+func newEncoder(w io.Writer, opts EncodeDecoderOptions) *encoder {
+	if opts == nil {
+		opts = NewEncodeDecoderOptions()
+	}
+	pool := opts.BytesPool()
+	return &encoder{
+		w:          bufio.NewWriterSize(w, opts.BufferSize()),
+		sizeBuffer: getByteSliceWithLength(sizeBufferSize, pool),
+		bytesPool:  pool,
+	}
+}
+
+func (e *encoder) Encode(m Marshaler) error {
+	size := m.Size()
+	if err := e.encodeSize(size); err != nil {
+		return err
+	}
+	return e.encodeData(m, size)
+}
+
+func (e *encoder) encodeSize(size int) error {
+	binary.BigEndian.PutUint32(e.sizeBuffer, uint32(size))
+	_, err := e.w.Write(e.sizeBuffer)
+	return err
+}
+
+func (e *encoder) encodeData(m Marshaler, size int) error {
+	e.dataBuffer = growDataBufferIfNeeded(e.dataBuffer, size, e.bytesPool)
+	size, err := m.MarshalTo(e.dataBuffer)
+	if err != nil {
+		return err
+	}
+	_, err = e.w.Write(e.dataBuffer[:size])
+	return err
+}
+
+func (e *encoder) resetWriter(w io.Writer) {
+	e.w.Reset(w)
+}
+
+func growDataBufferIfNeeded(buffer []byte, target int, pool pool.BytesPool) []byte {
```

Comment:
Pull this into a common file `common.go`? This is also used in `decode.go`.

Reply:
sounds good